plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java

```diff
@@ -250,6 +252,71 @@ public void testGetPartitionNames()
         assertEquals(mockClient.getAccessCount(), 2);
     }
 
+    /**
+     * Test {@link CachingHiveMetastore#getPartition(Table, List)} followed by
+     * {@link CachingHiveMetastore#getPartitionsByNames(Table, List)}.
+     * <p>
+     * At the moment of writing, CachingHiveMetastore uses HivePartitionName for keys in partition cache.
+     * HivePartitionName has a peculiar, semi- value-based equality. HivePartitionName may or may not be missing
+     * a name and it matters for bulk load, but it doesn't matter for single-partition load.
+     * Because of equality semantics, the cache keys may gets mixed during bulk load.
+     */
+    @Test
+    public void testGetPartitionThenGetPartitions()
+    {
+        Table table = metastore.getTable(TEST_DATABASE, TEST_TABLE).orElseThrow();
+        Optional<Partition> firstRead = metastore.getPartition(table, List.of(TEST_PARTITION1_VALUE));
+        assertThat(firstRead).isPresent();
+        assertThat(firstRead.get().getValues()).isEqualTo(List.of(TEST_PARTITION1_VALUE));
+
+        Map<String, Optional<Partition>> byName = metastore.getPartitionsByNames(table, List.of(TEST_PARTITION1));
+        assertThat(byName).containsOnlyKeys(TEST_PARTITION1);
+        Optional<Partition> secondRead = byName.get(TEST_PARTITION1);
+        assertThat(secondRead).isPresent();
+        assertThat(secondRead.get().getValues()).isEqualTo(List.of(TEST_PARTITION1_VALUE));
+    }
+
+    /**
+     * A variant of {@link #testGetPartitionThenGetPartitions} where the second get happens concurrently with eviction,
+     * here simulated with an explicit invalidation.
+     */
+    // Repeat test with invocationCount for better test coverage, since the tested aspect is inherently non-deterministic.
+    @Test(timeOut = 60_000, invocationCount = 20)
+    public void testGetPartitionThenGetPartitionsRacingWithInvalidation()
```

Comment:
Sadly, this didn't fail on CI. It fails very consistently locally.

Reply:
i have added another unit test that doesn't use threads and so is reliable. Keeping this one too for illustration purposes